### PR TITLE
Resolves LUIS v2 entities with roles

### LINKED ISSUE #166: Add roles support to LUIS v2
We should ensure that values with roles in LUIS v2 are picked up as expected.

### PR DESCRIPTION
**Pull request by rozele**
LUIS v3 sets the entity type for entities with roles to the role name. This change adds a similar resolution mechanism to LUIS v2, resolving entities with a "role" property to use the role for the entity type. This change also supports training LUIS v2 and v3 models with entities labeled with roles by adding the "roles" property to the `LuisSettings` supplied with the `--model-settings` CLI option.

Fixes #166

### CODE REVIEW

File: models/tests.json [hunk 1]
```diff
@@ -30,20 +30,20 @@
...
         "matchIndex": 0,
-        "entityValue": "2"
+        "entityValue": "5"
       }
     ]
   },
   {
     "text": "skip this one",
-    "intent": "Skip"
+    "intent": "Skip"
```
Comment (rozele) [inline models/tests.json]: Delete extra whitespace.

File: src/NLU.DevOps.Luis.Shared/LuisSettings.cs [hunk 2]
```diff
@@ -36,7 +36,7 @@ public LuisSettings(LuisApp appTemplate)
         /// </summary>
         /// <param name="prebuiltEntityTypes">Prebuilt entity types.</param>
         public LuisSettings(IReadOnlyDictionary<string, string> prebuiltEntityTypes)
-            : this(null, prebuiltEntityTypes)
+            : this(null, prebuiltEntityTypes, null)
```
Comment (saroup) [inline src/NLU.DevOps.Luis.Shared/LuisSettings.cs]: >prebuiltEntityTypes [](start = 25, length = 19)

you don't want to add a check to see if this is null and create an empty dict instead? Same for appTemplate above?

File: src/NLU.DevOps.Luis.Shared/LabeledUtteranceExtensions.cs [hunk 3]
```diff
@@ -4,22 +4,27 @@
...
 
     internal static class LabeledUtteranceExtensions
     {
-        public static JSONUtterance ToJSONUtterance(this Models.LabeledUtterance utterance, IReadOnlyDictionary<string, string> prebuiltEntityTypes)
+        public static JSONUtterance ToJSONUtterance(this Models.LabeledUtterance utterance, LuisSettings luisSettings)
         {
             JSONEntity toJSONEntity(Entity entity)
             {
-                var entityType = prebuiltEntityTypes.TryGetValue(entity.EntityType, out var builtinType)
+                var startPos = entity.StartCharIndexInText(utterance.Text);
```
Comment (saroup) [inline src/NLU.DevOps.Luis.Shared/LabeledUtteranceExtensions.cs]: >entity [](start = 31, length = 6)

risk of null entity here?

File: models/settings.luis.json [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
+    "songCount": "number"
   },
   "appTemplate": {
     "prebuiltEntities": [
       {
-        "name": "number"
+        "name": "number",
+        "roles": [
+          "songCount"
+        ]
```
Comment (saroup) [inline models/settings.luis.json]: Do we have to add documentation for this change?

File: src/NLU.DevOps.Luis.Shared/LuisSettings.cs (hunk 2 above)
Comment (rozele) [reply src/NLU.DevOps.Luis.Shared/LuisSettings.cs]: null check occurs in the constructor below.

File: src/NLU.DevOps.Luis.Shared/LabeledUtteranceExtensions.cs (hunk 3 above)
Comment (rozele) [reply src/NLU.DevOps.Luis.Shared/LabeledUtteranceExtensions.cs]: This is an internal class, and LUIS doesn't respond with null entities.

File: models/settings.luis.json (hunk 4 above)
Comment (rozele) [reply models/settings.luis.json]: Yes.

Comment (rozele) [reply models/settings.luis.json]: Done.

**Pull request merged**